src/develop/blend.c

```diff
@@ -513,6 +577,172 @@ void dt_develop_blend_process(struct dt_iop_module_t *self, struct dt_dev_pixelp
+  float *lum = NULL;
+  cl_mem tmp = NULL;
+  cl_mem blur = NULL;
+  cl_mem out = NULL;
+
+  dt_dev_pixelpipe_t *p = piece->pipe;
+  if(p->luminance_mask_data == NULL) goto error;
+
+  const int iwidth  = p->luminance_mask_roi.width;
+  const int iheight = p->luminance_mask_roi.height;
+  const int owidth  = roi_in->width;
+  const int oheight = roi_in->height;
+
+  lum = dt_alloc_align_float(iwidth * iheight);
+  if(lum == NULL) goto error;
+  tmp = dt_opencl_alloc_device(devid, iwidth, iheight, sizeof(float));
+  if(tmp == NULL) goto error;
+  out = dt_opencl_alloc_device_buffer(devid, iwidth * iheight * sizeof(float));
+  if(out == NULL) goto error;
+  blur = dt_opencl_alloc_device_buffer(devid, iwidth * iheight * sizeof(float));
+  if(blur == NULL) goto error;
+
+  {
+    const int err = dt_opencl_write_host_to_device(devid, p->luminance_mask_data, tmp, iwidth, iheight, sizeof(float));
+    if(err != CL_SUCCESS) goto error;
+  }
+
+  const int program = 31;
+  int kernel_write_mask = dt_opencl_create_kernel(program, "writeout_mask");
+  int kernel_read_mask  = dt_opencl_create_kernel(program, "readin_mask");
+  int kernel_calc_blend = dt_opencl_create_kernel(program, "calc_detail_blend");
+  int kernel_mask_blur  = dt_opencl_create_kernel(program, "fastblur_mask_9x9");
+
+  {
+    size_t sizes[3] = { ROUNDUPWD(iwidth), ROUNDUPHT(iheight), 1 };
+    dt_opencl_set_kernel_arg(devid, kernel_read_mask, 0, sizeof(cl_mem), &out);
+    dt_opencl_set_kernel_arg(devid, kernel_read_mask, 1, sizeof(cl_mem), &tmp);
+    dt_opencl_set_kernel_arg(devid, kernel_read_mask, 2, sizeof(int), &iwidth);
+    dt_opencl_set_kernel_arg(devid, kernel_read_mask, 3, sizeof(int), &iheight);
+    const int err = dt_opencl_enqueue_kernel_2d(devid, kernel_read_mask, sizes);
+    if(err != CL_SUCCESS) goto error;
+  }  
+
+  {
+    size_t sizes[3] = { ROUNDUPWD(iwidth), ROUNDUPHT(iheight), 1 };
+    dt_opencl_set_kernel_arg(devid, kernel_calc_blend, 0, sizeof(cl_mem), &out);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_blend, 1, sizeof(cl_mem), &blur);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_blend, 2, sizeof(int), &iwidth);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_blend, 3, sizeof(int), &iheight);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_blend, 4, sizeof(float), &threshold);
+    dt_opencl_set_kernel_arg(devid, kernel_calc_blend, 5, sizeof(int), &detail);
+    const int err = dt_opencl_enqueue_kernel_2d(devid, kernel_calc_blend, sizes);
+    if(err != CL_SUCCESS) goto error;
+  }  
+
+  {
+    // For a blurring sigma of 2.0f a 13x13 kernel would be optimally required but the 9x9 is by far good enough here 
+    float kernel[9][9];
+    const double temp = -2.0f * 2.0f * 2.0f;
+    float sum = 0.0f;
+    for(int i = -4; i <= 4; i++)
+    {
+      for(int j = -4; j <= 4; j++)
+      {
+        kernel[i + 4][j + 4] = expf( ((i*i) + (j*j)) / temp);
+        sum += kernel[i + 4][j + 4];
+      }
+    }
+    for(int i = 0; i < 9; i++)
```

Comment:
do you think the compiler could generate SIMD code here?

Reply:
Yes although the benefit won't be measurable :-)

BTW there is a huge amount of work done mask-wise. I still feel quite bad about "how-to-git" this pr correctly.
1. git rebase master (maybe i'll have to resolve conflicts)
2. do i have to push / force push then before continuing work?

I already have messed up pr's somehow and had to start again via "meld" ... Could you give a hint?

Reply:
maybe better use
```
#if defined(__GNUC__)
  #pragma GCC ivdep
#endif
```
inside the loop here

Reply:
> Yes although the benefit won't be measurable :-)
> 
> BTW there has been a huge amount of work done mask-wise in other pr's. I still feel quite bad about "how-to-git" this pr correctly.
> 
>     1. git rebase master (maybe i'll have to resolve conflicts)
> 
>     2. do i have to push / force push then before continuing work?
> 
> 
> I already have messed up pr's somehow and had to start again via "meld" ... Could you give a hint?

For a topic branch like this one, you need to rebase, resolve conflicts and force-push on your branch. Doing that you can also clean-up the commits by merging some which are not useful for the final merge.

Reply:
> maybe better use
> 
> ```
> #if defined(__GNUC__)
>   #pragma GCC ivdep
> #endif
> ```
> 
> inside the loop here

Maybe, I'm not expert here!